addons/udes_stock_routing/models/stock_move.py

```diff
@@ -105,15 +106,165 @@ def _get_new_picking_values(self):
+
+        param: assigned_moves: stock.move(x,) recordset.
+        param: rule: stock.rule(x,) singleton.
+        """
+        StockMoveObj = self.env["stock.move"]
+        res = StockMoveObj.browse()
+        # Rip the moves to the operation type of the rule.
+        split_pick_vals = {
+            "location_id": rule.location_src_id.id,
+            "location_dest_id": rule.location_id.id,
+            "picking_type_id": rule.picking_type_id.id,
+            "origin": assigned_moves.picking_id.origin,
+        }
+
+        # We only need to split the moves to a backorder if the picking type
+        # of the rule differs from the assigned moves picking type. If it
+        # is the same picking type, any post assign behaviour has already run, and
+        # the move lines and moves already exist on the picking of the correct type.
+        if assigned_moves.picking_type_id != rule.picking_type_id:
+            split_pick = assigned_moves.picking_id._create_backorder_picking(assigned_moves, **split_pick_vals)
+            # Ensure any candidate moves are merged, because the new picking type could have
+            # different configuration and want to handle them differently. Get back to square 1 basically.
+            split_pick.move_lines._merge_moves()
+            # Update the split moves picking type to the rules picking type, to ensure
+            # that when _action_assign() is called, the applicable rules get detected correctly.
+            split_pick.move_lines.write(
+                {
+                    "picking_type_id": rule.picking_type_id.id,
+                    "location_id": rule.location_src_id.id,
+                }
+            )
+            res |= split_pick.move_lines
+            # Re-trigger _action_assign() on the new split moves as they got split
+            # to a new picking type and may need to run other post assign action behaviour.
+            split_pick.move_lines._action_assign()
+
+        return res
+
     def _action_assign(self):
         """
-        Extend action_assign to include the two stage hook.
+        Extend action_assign to include the two stage hook and split pick hook.
         This is done at the move level incase refactoring splits picks after assigning stock.
         This way, we don't lose any moves which need to be split to 2 stage but got refactored out.
         """
-        res = super()._action_assign()
+        StockMoveObj = self.env["stock.move"]
+        Rule = self.env["stock.rule"]
+
+        res = StockMoveObj.browse()
+        applicable_rules = Rule.browse()
+        # Allows us to build data on which moves have been split for which rules
+        # which we can then split to backorder pickings after looping over all rules.
+        moves_to_split_by_rule = collections.defaultdict(StockMoveObj.browse)
+        applicable_rules = self.get_split_pick_applicable_rules()
+
+        if not applicable_rules:
```

Comment:
We have moved this up but we still executing all the code, so we either add an else and indent or we call a function to do all the logic from lines 172 til 200.

Reply:
I didnt notice, but i think should be fine just to return if not applicable_rules?

Reply:
nope, we need to run the two stage hook afterwards

Reply:
it is hopefully a little easier to read now anyway.